# Livro para consulta:
- https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html
- https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html
    

# 1. Importando bibliotecas <a name="import"></a>

<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# 2. Carregando o dataframe SINASC <a name="read"></a>
<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [17]:
sinasc_raw = pd.read_csv('SINASC_RO_2019.csv')


# Tarefa 1

### 1. Idade media das mães e dos pais por município (coluna munResNome)


In [18]:
sinasc_1= sinasc_raw[['munResNome', 'IDADEMAE', 'IDADEPAI']]

In [19]:
sinasc_1.groupby('munResNome').mean()

IDADEMAE   IDADEPAI
munResNome                                     
Alta Floresta D'Oeste      25.991826  29.548701
Alto Alegre dos Parecis    24.844156  29.179104
Alto Paraíso               24.960177  28.833333
Alvorada D'Oeste           25.770732  30.757282
Ariquemes                  25.607866  32.466667
Buritis                    25.559177  30.812500
Cabixi                     26.037500  34.333333
Cacaulândia                25.546667  36.333333
Cacoal                     26.890102  30.931854
Campo Novo de Rondônia     24.825301  30.500000
Candeias do Jamari         25.243323  29.894737
Castanheiras               27.281250  30.392857
Cerejeiras                 27.240506  31.652778
Chupinguaia                25.333333  29.629630
Colorado do Oeste          27.602804  34.354839
Corumbiara                 24.716535  32.470588
Costa Marques              24.412017  30.613636
Cujubim                    24.482927  31.400000
Espigão D'Oeste            26.093333  31.198113
Governador Jorge Teixeira  24.401786  29.222222
Guajará-Mirim              25.462228  30.470588
Itapuã do Oeste            26.064220  42.500000
Jaru                       25.982974  32.197183
Ji-Paraná                  26.361137  31.132239
Machadinho D'Oeste         25.480000  31.326923
Ministro Andreazza         27.904762  33.849315
Mirante da Serra           26.304878  32.154762
Monte Negro                25.648241  34.678571
Município ignorado - RO    24.000000  22.000000
Nova Brasilândia D'Oeste   26.222222  30.520833
Nova Mamoré                25.326389  31.692308
Nova União                 26.894231  31.371429
Novo Horizonte do Oeste    26.938776  31.588235
Ouro Preto do Oeste        26.375594  31.233244
Parecis                    26.454545  32.473684
Pimenta Bueno              25.199675  29.745174
Pimenteiras do Oeste       25.075000  35.600000
Porto Velho                26.312670  32.352679
Presidente Médici          25.796053  30.150754
Primavera de Rondônia      27.023256  31.888889
Rio Crespo                 26.660000  28.000000
Rolim de Moura             26.600863  30.360479
Santa Luzia D'Oeste        26.121212  30.703297
Seringueiras               24.910714  30.500000
São Felipe D'Oeste         26.481481  31.395833
São Francisco do Guaporé   25.660066  31.638889
São Miguel do Guaporé      25.926346  32.666667
Teixeirópolis              26.031250  31.500000
Theobroma                  25.536082  30.083333
Urupá                      26.688235  31.008929
Vale do Anari              26.096774  34.000000
Vale do Paraíso            25.679012  29.976744
Vilhena                    26.433962  34.152941

Logo acima, é possivel concluir que mães são mais jovens que os pais, independente de qual municipio, nenhum tem o cenário inverso.
Abaixo, podemos verificar um quadro mais complexo, e ai concluimos que homens tem a idade média próximo aos 30 anos, enquanto mulheres, ficam na casa de 24 anos em diante.

In [7]:
agrupamento = sinasc_1.groupby(['munResNome'])
agrupamento.agg(['mean', 'std', amplitude],ascending=True)

IDADEMAE                       IDADEPAI  \
                                mean       std amplitude       mean   
munResNome                                                            
Alta Floresta D'Oeste      25.991826  6.369535        39  29.548701   
Alto Alegre dos Parecis    24.844156  5.957507        27  29.179104   
Alto Paraíso               24.960177  6.211331        31  28.833333   
Alvorada D'Oeste           25.770732  6.442135        29  30.757282   
Ariquemes                  25.607866  6.193641        31  32.466667   
Buritis                    25.559177  6.142594        30  30.812500   
Cabixi                     26.037500  6.212368        26  34.333333   
Cacaulândia                25.546667  6.226129        26  36.333333   
Cacoal                     26.890102  6.295376        33  30.931854   
Campo Novo de Rondônia     24.825301  5.357483        27  30.500000   
Candeias do Jamari         25.243323  6.314677        31  29.894737   
Castanheiras               27.281250  5.335423        22  30.392857   
Cerejeiras                 27.240506  6.281292        38  31.652778   
Chupinguaia                25.333333  6.461544        29  29.629630   
Colorado do Oeste          27.602804  6.615261        32  34.354839   
Corumbiara                 24.716535  5.771987        28  32.470588   
Costa Marques              24.412017  6.656099        29  30.613636   
Cujubim                    24.482927  5.901641        28  31.400000   
Espigão D'Oeste            26.093333  6.430731        30  31.198113   
Governador Jorge Teixeira  24.401786  5.698838        28  29.222222   
Guajará-Mirim              25.462228  6.886640        30  30.470588   
Itapuã do Oeste            26.064220  7.005653        31  42.500000   
Jaru                       25.982974  6.174429        33  32.197183   
Ji-Paraná                  26.361137  6.220517        32  31.132239   
Machadinho D'Oeste         25.480000  6.435560        31  31.326923   
Ministro Andreazza         27.904762  6.223970        27  33.849315   
Mirante da Serra           26.304878  6.342357        32  32.154762   
Monte Negro                25.648241  6.284250        26  34.678571   
Município ignorado - RO    24.000000       NaN         0  22.000000   
Nova Brasilândia D'Oeste   26.222222  5.777646        36  30.520833   
Nova Mamoré                25.326389  6.478768        32  31.692308   
Nova União                 26.894231  5.636080        23  31.371429   
Novo Horizonte do Oeste    26.938776  5.687418        26  31.588235   
Ouro Preto do Oeste        26.375594  6.659518        30  31.233244   
Parecis                    26.454545  6.489543        25  32.473684   
Pimenta Bueno              25.199675  6.151005        34  29.745174   
Pimenteiras do Oeste       25.075000  6.289542        26  35.600000   
Porto Velho                26.312670  6.549074        35  32.352679   
Presidente Médici          25.796053  6.523215        30  30.150754   
Primavera de Rondônia      27.023256  5.974104        23  31.888889   
Rio Crespo                 26.660000  6.129920        23  28.000000   
Rolim de Moura             26.600863  6.407535        30  30.360479   
Santa Luzia D'Oeste        26.121212  5.962934        26  30.703297   
Seringueiras               24.910714  6.063368        28  30.500000   
São Felipe D'Oeste         26.481481  6.308846        24  31.395833   
São Francisco do Guaporé   25.660066  6.078402        27  31.638889   
São Miguel do Guaporé      25.926346  5.851400        28  32.666667   
Teixeirópolis              26.031250  5.631456        23  31.500000   
Theobroma                  25.536082  6.364062        27  30.083333   
Urupá                      26.688235  5.913817        28  31.008929   
Vale do Anari              26.096774  6.324666        30  34.000000   
Vale do Paraíso            25.679012  6.974645        38  29.976744   
Vilhena                    26.433962  6.310937        32  34.152941   

                                                
  

### 2. Peso médio dos bebes por sexo que nasceram no dia do seu aniversário por faixas de escolaridade mae
Ex: Você, aluna(o), nasceu no dia 10/01, então você precisa filtrar o conjunto de dados nessa data e calcular o peso médio dos bebês de cada sexo por faixa de escolaridade da mãe.

In [8]:
sinasc_raw.DTNASC = pd.to_datetime(sinasc_raw.DTNASC)
sinasc_raw.DTNASC

0       2019-02-19
1       2019-02-21
2       2019-02-25
3       2019-03-20
4       2019-03-23
           ...    
27023   2019-12-13
27024   2019-10-04
27025   2019-08-02
27026   2019-12-23
27027   2019-12-27
Name: DTNASC, Length: 27028, dtype: datetime64[ns]

In [9]:
sinasc_aniversario = sinasc_raw[(sinasc_raw.DTNASC.dt.day == 5) & 
                                (sinasc_raw.DTNASC.dt.month == 7)]

sinasc_aniversario = sinasc_aniversario.groupby(by=['DTNASC',
                                                    'SEXO',
                                                    'ESCMAE'])['PESO'].mean()

sinasc_aniversario = sinasc_aniversario.to_frame().reindex(labels=['1 a 3 anos', 
                                                                   '4 a 7 anos', 
                                                                   '8 a 11 anos', 
                                                                   '12 anos ou mais'], 
                                                           level='ESCMAE')

sinasc_aniversario

PESO
DTNASC     SEXO      ESCMAE                      
2019-07-05 Feminino  4 a 7 anos       3269.500000
                     8 a 11 anos      3376.875000
                     12 anos ou mais  3099.166667
           Masculino 1 a 3 anos       3050.000000
                     4 a 7 anos       2990.000000
                     8 a 11 anos      3325.000000
                     12 anos ou mais  2875.500000

A minha data de nascimento é 05 de julho, portanto, podemos verificar que nesse dia as meninas, nasceram um pouco mais pesadas, outro fato interessante é que, neste exato dia, mulheres com estudo de 1 a 3 anos, tiveram apenas bebês do sexo masculino.

### 3. Qual o municipio que nasceu menos bebe em 2019?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

O municipio que teve a menor taxa de nascimento é Castanheiras, com a idade média da mãe de 27 anos e do pai, 30 anos.
A faixa de idade minima e maxima da mãe são respectivamente os valores de, 39 e 17 anos. Já o pai, minima e máxima são, 17 e 43 anos.

In [26]:
sinasc_raw['COUNT'] = 1
sinasc_3 = sinasc_raw.groupby(by='munResNome')[['COUNT', 
                                                'IDADEMAE',
                                                'IDADEPAI']].agg(
    {'COUNT':[('Quantidade de bebês nascidos','count')], 
     'IDADEMAE':[('idade média','mean'), 
                 ('idade máxima',max), 
                 ('idade mínima',min)], 
     'IDADEPAI':[('idade média','mean'), 
                 ('idade máxima',max), 
                 ('idade mínima',min)]})

sinasc_3.sort_values(('COUNT', 'Quantidade de bebês nascidos'))

COUNT    IDADEMAE  \
                          Quantidade de bebês nascidos idade média   
munResNome                                                           
Município ignorado - RO                              1   24.000000   
Castanheiras                                        32   27.281250   
Pimenteiras do Oeste                                40   25.075000   
Primavera de Rondônia                               43   27.023256   
Parecis                                             44   26.454545   
Rio Crespo                                          50   26.660000   
São Felipe D'Oeste                                  54   26.481481   
Teixeirópolis                                       64   26.031250   
Cacaulândia                                         75   25.546667   
Cabixi                                              80   26.037500   
Vale do Paraíso                                     81   25.679012   
Ministro Andreazza                                  84   27.904762   
Vale do Anari                                       93   26.096774   
Theobroma                                           97   25.536082   
Novo Horizonte do Oeste                             98   26.938776   
Santa Luzia D'Oeste                                 99   26.121212   
Nova União                                         104   26.894231   
Itapuã do Oeste                                    109   26.064220   
Governador Jorge Teixeira                          112   24.401786   
Corumbiara                                         127   24.716535   
Alto Alegre dos Parecis                            154   24.844156   
Chupinguaia                                        162   25.333333   
Mirante da Serra                                   164   26.304878   
Campo Novo de Rondônia                             166   24.825301   
Seringueiras                                       168   24.910714   
Urupá                                              170   26.688235   
Monte Negro                                        199   25.648241   
Cujubim                                            205   24.482927   
Alvorada D'Oeste                                   205   25.770732   
Colorado do Oeste                                  214   27.602804   
Alto Paraíso                                       226   24.960177   
Costa Marques                                      233   24.412017   
Cerejeiras                                         237   27.240506   
Nova Brasilândia D'Oeste                           261   26.222222   
São Francisco do Guaporé                           303   25.660066   
Presidente Médici                                  304   25.796053   
Candeias do Jamari                                 337   25.243323   
São Miguel do Guaporé                              353   25.926346   
Alta Floresta D'Oeste                              367   25.991826   
Nova Mamoré                                        432   25.326389   
Espigão D'Oeste                                    450   26.093333   
Machadinho D'Oeste                                 500   25.480000   
Buritis                                            583   25.559177   
Pimenta Bueno                                      616   25.199675   
Ouro Preto do Oeste                                631   26.375594   
Guajará-Mirim                                      781   25.462228   
Jaru                                               881   25.982974   
Rolim de Moura                                     927   26.600863   
Cacoal                                            1374   26.890102   
Vilhena                                           1590   26.433962   
Ariquemes                                         1729   25.607866   
Ji-Paraná                                         2182   26.361137   
Porto Velho                                       8437   26.312670   

                                                       IDADEPAI               \
                          idade máxima idade mín

### 4. Qual o municipio que nasceu mais bebe no mês de março?
    - qual a quantidade de filhos vivos media, maxima, minima nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?



In [11]:
sinasc_mar = sinasc_raw[sinasc_raw.DTNASC.dt.month == 3].copy()

In [20]:
sinasc_mar['MES'] = sinasc_mar['DTNASC'].dt.month_name()
sinasc_mar['COUNT'] = 1
sinasc_4 = sinasc_mar.groupby(by=['MES',
                                  'munResNome']
                             )[['COUNT', 
                                'QTDFILVIVO',
                                'IDADEMAE',
                                'IDADEPAI']
                              ].agg({'COUNT':[('Qtd de bebês nascidos','count')], 
                                     'QTDFILVIVO':[('Média','mean'), 
                                                   ('Máx',max), 
                                                   ('Mín',min)], 
                                     'IDADEMAE':[('Média','mean'), 
                                                 ('Máx',max), 
                                                 ('Mín',min), 
                                                 ('Amplitude',lambda x: x.max()-x.min())], 
                                     'IDADEPAI':[('Média','mean'), 
                                                 ('Máx',max), 
                                                 ('Mín',min), 
                                                 ('Amplitude',lambda x: x.max()-x.min())]
                                    })

sinasc_4['Diferença da média IDADEPAI-IDADEMAE'] = sinasc_4[('IDADEPAI','Média')]-sinasc_4[('IDADEMAE','Média')]
sinasc_4.sort_values(('COUNT', 'Qtd de bebês nascidos'), 
                     ascending=False
                    )

COUNT QTDFILVIVO            \
                                Qtd de bebês nascidos      Média  Máx  Mín   
MES   munResNome                                                             
March Porto Velho                                 744   1.295056  9.0  0.0   
      Ji-Paraná                                   188   0.828877  4.0  0.0   
      Vilhena                                     148   0.979592  5.0  0.0   
      Ariquemes                                   141   1.064748  4.0  0.0   
      Cacoal                                      133   0.969697  7.0  0.0   
      Jaru                                         80   1.025641  5.0  0.0   
      Guajará-Mirim                                75   1.534247  8.0  0.0   
      Rolim de Moura                               69   0.724638  5.0  0.0   
      Ouro Preto do Oeste                          67   1.014925  5.0  0.0   
      Pimenta Bueno                                62   0.836066  4.0  0.0   
      Buritis                                      59   1.421053  4.0  0.0   
      Nova Mamoré                                  53   1.360000  6.0  0.0   
      Espigão D'Oeste                              44   1.386364  6.0  0.0   
      Machadinho D'Oeste                           44   1.000000  6.0  0.0   
      São Miguel do Guaporé                        41   1.000000  4.0  0.0   
      Candeias do Jamari                           33   1.250000  4.0  0.0   
      São Francisco do Guaporé                     31   1.161290  6.0  0.0   
      Alta Floresta D'Oeste                        31   0.903226  4.0  0.0   
      Presidente Médici                            29   0.724138  3.0  0.0   
      Costa Marques                                24   1.208333  4.0  0.0   
      Alvorada D'Oeste                             22   1.000000  3.0  0.0   
      Alto Paraíso                                 22   1.285714  3.0  0.0   
      Monte Negro                                  20   0.700000  3.0  0.0   
      Nova Brasilândia D'Oeste                     20   0.947368  3.0  0.0   
      Colorado do Oeste                            19   1.000000  4.0  0.0   
      Seringueiras                                 19   1.000000  2.0  0.0   
      Chupinguaia                                  18   1.055556  4.0  0.0   
      Cujubim                                      17   1.294118  5.0  0.0   
      Cerejeiras                                   16   0.875000  2.0  0.0   
      Urupá                                        14   0.928571  2.0  0.0   
      Vale do Paraíso                              13   0.583333  2.0  0.0   
      Campo Novo de Rondônia                       13   1.285714  2.0  0.0   
      Corumbiara                                   12   0.909091  3.0  0.0   
      Vale do Anari                                11   0.818182  2.0  0.0   
      Alto Alegre dos Parecis                      10   0.888889  2.0  0.0   
      Governador Jorge Teixeira                    10   0.555556  2.0  0.0   
      Cacaulândia                                  10   1.200000  4.0  0.0   
      Rio Crespo                                    9   1.000000  3.0  0.0   
      Mirante da Serra                              9   1.000000  2.0  0.0   
      Theobroma                                     8   1.125000  2.0  0.0   
      Itapuã do Oeste                               8   1.285714  3.0  0.0   
      Santa Luzia D'Oeste                           7   0.428571  2.0  0.0   
      Nova União                                    7   0.571429  1.0  0.0   
      Cabixi                                        6   0.833333  3.0  0.0   
      Parecis                                       6   1.833333  5.0  1.0   
      Ministro Andreazza                            6   1.166667  3.0  0.0   
      Pimenteiras do Oeste                          6   1.000000  2.0  0.0   
      Primavera de Rondônia                         5   1.200000  2.0  0.0   
      São Felipe D'Oeste                            5   0.000000  0.0 

No quadro acima, podemos verificar que o municipio com a maior taxa de nascimento para o mês de março é Porto Velho , com 744 bebês. A idade média da mãe é de 26,6 anos e do pai, 34 anos, ainda falando sobre o pai, a idade minima que encontramos é de 19 e a máxima é de 62 anos. A quantidade de filhos vivos média é de 1 e a máxima é de 9 crianças.


### Analise as respostas encontradas, tire algum insight delas, conte pra gente algo encontrado nos dados. Algo que você julgue relevante e novo pra você.

Exemplo:
- Ah, descobri que a idade mediana das mulheres que deram a luz no ano de 2019 dos municipios x é maior que y.

In [13]:
sinasc_raw.dtypes

ORIGEM          int64
CODESTAB      float64
CODMUNNASC      int64
LOCNASC         int64
IDADEMAE        int64
               ...   
munResUf       object
munResLat     float64
munResLon     float64
munResAlt     float64
munResArea    float64
Length: 69, dtype: object

In [14]:
sinasc_raw.select_dtypes(exclude=['int64','float64'])

ESCMAE         GESTACAO GRAVIDEZ    PARTO     DTNASC  \
0          8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-02-19   
1          8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-02-21   
2          8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-02-25   
3      12 anos ou mais  37 a 41 semanas    Única  Cesáreo 2019-03-20   
4          8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-03-23   
...                ...              ...      ...      ...        ...   
27023  12 anos ou mais  32 a 36 semanas    Única  Cesáreo 2019-12-13   
27024      8 a 11 anos  37 a 41 semanas    Única  Vaginal 2019-10-04   
27025      8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-08-02   
27026      8 a 11 anos  32 a 36 semanas    Única  Vaginal 2019-12-23   
27027      8 a 11 anos  37 a 41 semanas    Única  Cesáreo 2019-12-27   

            SEXO RACACOR  DTCADASTRO CODANOMAL VERSAOSIST   DTRECEBIM  \
0      Masculino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
1       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
2       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
3       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
4       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
...          ...     ...         ...       ...        ...         ...   
27023  Masculino  Branca  2020-09-28       NaN     3.2.01  2020-10-13   
27024  Masculino  Branca  2019-10-16       NaN     3.2.01  2019-10-21   
27025  Masculino   Parda  2019-08-07       NaN     3.2.00  2019-08-07   
27026  Masculino   Parda  2020-01-07       NaN     3.2.00  2020-01-07   
27027  Masculino   Parda  2020-10-15       NaN     3.2.01  2020-10-16   

      munResStatus munResTipo               munResNome  munResUf  
0            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
1            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
2            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
3            ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia  
4            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
...            ...        ...                      ...       ...  
27023        ATIVO      MUNIC                  Vilhena  Rondônia  
27024        ATIVO      MUNIC              Chupinguaia  Rondônia  
27025        ATIVO      MUNIC                  Vilhena  Rondônia  
27026        ATIVO      MUNIC                  Vilhena  Rondônia  
27027        ATIVO      MUNIC  Novo Horizonte do Oeste  Rondônia  

[27028 rows x 15 columns]

In [15]:
sinasc_raw.columns

Index(['ORIGEM', 'CODESTAB', 'CODMUNNASC', 'LOCNASC', 'IDADEMAE', 'ESTCIVMAE',
       'ESCMAE', 'CODOCUPMAE', 'QTDFILVIVO', 'QTDFILMORT', 'CODMUNRES',
       'GESTACAO', 'GRAVIDEZ', 'PARTO', 'CONSULTAS', 'DTNASC', 'HORANASC',
       'SEXO', 'APGAR1', 'APGAR5', 'RACACOR', 'PESO', 'IDANOMAL', 'DTCADASTRO',
       'CODANOMAL', 'NUMEROLOTE', 'VERSAOSIST', 'DTRECEBIM', 'DIFDATA',
       'DTRECORIGA', 'NATURALMAE', 'CODMUNNATU', 'CODUFNATU', 'ESCMAE2010',
       'SERIESCMAE', 'DTNASCMAE', 'RACACORMAE', 'QTDGESTANT', 'QTDPARTNOR',
       'QTDPARTCES', 'IDADEPAI', 'DTULTMENST', 'SEMAGESTAC', 'TPMETESTIM',
       'CONSPRENAT', 'MESPRENAT', 'TPAPRESENT', 'STTRABPART', 'STCESPARTO',
       'TPNASCASSI', 'TPFUNCRESP', 'TPDOCRESP', 'DTDECLARAC', 'ESCMAEAGR1',
       'STDNEPIDEM', 'STDNNOVA', 'CODPAISRES', 'TPROBSON', 'PARIDADE',
       'KOTELCHUCK', 'CONTADOR', 'munResStatus', 'munResTipo', 'munResNome',
       'munResUf', 'munResLat', 'munResLon', 'munResAlt', 'munResArea'],
      dtype='object')